src/main/java/com/amazon/ion/impl/Ion11Writer.kt

```diff
@@ -0,0 +1,153 @@
+
+    /** Writes one annotation for the next value. */
+    fun writeAnnotations(annotation0: String)
+
+    /** Writes two annotations for the next value. */
+    fun writeAnnotations(annotation0: String, annotation1: String)
+
+    /** Writes three or more annotations for the next value. */
+    fun writeAnnotations(annotation0: String, annotation1: String, vararg annotations: String)
+
+    /**
+     * Writes the field name for the next value. Must be called while in a struct and must be called before [writeAnnotations].
+     * @throws com.amazon.ion.IonException if annotations are already written for the value or if not in a struct.
+     */
+    fun writeFieldName(text: String)
+
+    /**
+     * Writes the field name for the next value. Must be called while in a struct and must be called before [writeAnnotations].
+     * @throws com.amazon.ion.IonException if annotations are already written for the value or if not in a struct.
+     */
+    fun writeFieldName(sid: Long)
+
+    fun stepInList(delimited: Boolean)
+    fun stepInSExp(delimited: Boolean)
+
+    /**
+     * Delimited struct with FlexSym field names
+     * Variable length struct with symbol address field names
+     * Variable length struct with FlexSym field names
+     *
+     * @throws com.amazon.ion.IonException if delimited is true and useFlexSym is false.
+     */
+    fun stepInStruct(delimited: Boolean, useFlexSym: Boolean)
+
+    /**
+     * Steps into a stream.
+     * A stream is not a container in the Ion data model, but it is a container from an encoding perspective.
+     */
+    fun stepInStream()
+
+    /**
+     * Writes a macro invocation for the given macro name.
+     * A macro is not a container in the Ion data model, but it is a container from an encoding perspective.
+     */
+    fun stepInEExp(name: String)
+
+    /**
+     * Writes a macro invocation for the given id corresponding to a macro in the macro table.
+     * A macro is not a container in the Ion data model, but it is a container from an encoding perspective.
+     */
+    fun stepInEExp(id: Long)
+
+    /**
+     * Steps out of the current container.
+     */
+    fun stepOut()
+
+    // TODO: Doc comments for the uninteresting functions
+
+    fun writeNull()
+    fun writeNull(type: IonType)
+
+    fun writeBool(value: Boolean)
+
+    fun writeInt(value: Byte)
+    fun writeInt(value: Short)
+    fun writeInt(value: Int)
+    fun writeInt(value: Long)
+    fun writeInt(value: BigInteger)
+
+    fun writeFloat(value: Float)
+    fun writeFloat(value: Double)
+
+    fun writeDecimal(value: BigDecimal)
+    fun writeDecimal(value: Decimal)
+
+    fun writeTimestamp(value: Timestamp)
+
+    /** Writes an Ion timestamp with Days precision */
+    fun writeTimestamp(value: LocalDate)
+
+    /** Writes an Ion timestamp with 0 offset (UTC) and nanosecond precision. */
+    fun writeTimestamp(value: Instant)
+
+    fun writeSymbol(id: Long)
+    fun writeSymbol(text: CharSequence)
```

Comment:
We use String some places and CharSequence in others here. What's the rationale for each?

When is this written inline vs as a symbol ID?

Reply:
The String vs CharSequence issue is something I need to clean up. No rationale for having both.

Re. When inline be symbol ID?
That's an implementation detail. The raw writer will write the string arg as an inline symbol, but a "managed" writer could choose to intern the symbol text and write a numeric ID.

Reply:
I don't think it's quite that simple, because (as with all symbol tokens) the *user* may want to specify that some symbols should be inlined rather than interned. Or, we need to provide a detection mechanism with some configurable options to allow the user to influence when things should be inlined vs. interned. Is that what you had in mind?

Reply:
Yes, what I had in mind is configuration driven "managed" writers. I say writers (plural) because I think we''ll want managed writers for both text and binary so that the text writer will also be able to write out templates.

I suspect that users who care about how individual symbols are encoded will probably want to manage their own symbol table too, so they can just use the raw writer, which would give them that control.

Reply:
>I suspect that users who care about how individual symbols are encoded will probably want to manage their own symbol table too

I don't think that's necessarily true, but I'm happy to move forward for now.